pkg/gpu/testutil/mocks.go

```diff
@@ -130,84 +155,118 @@ func GetDeviceMock(deviceIdx int) *nvmlmock.Device {
 		GetIndexFunc: func() (int, nvml.Return) {
 			return deviceIdx, nvml.SUCCESS
 		},
+		IsMigDeviceHandleFunc: func() (bool, nvml.Return) {
+			return false, nvml.SUCCESS
+		},
 	}
-}
 
-// GetBasicNvmlMock returns a mock of the nvml.Interface with a single device with 10 cores,
-// useful for basic tests that need only the basic interaction with NVML to be working.
-func GetBasicNvmlMock() *nvmlmock.Interface {
-	if len(GPUUUIDs) != len(GPUCores) {
-		// Make it really easy to spot errors if we change any of the arrays.
-		panic("GPUUUIDs and GPUCores must have the same length, please fix it")
+	for _, opt := range opts {
+		opt(mock)
 	}
 
-	return &nvmlmock.Interface{
-		DeviceGetCountFunc: func() (int, nvml.Return) {
-			return len(GPUUUIDs), nvml.SUCCESS
-		},
-		DeviceGetHandleByIndexFunc: func(index int) (nvml.Device, nvml.Return) {
-			return GetDeviceMock(index), nvml.SUCCESS
-		},
-		DeviceGetCudaComputeCapabilityFunc: func(nvml.Device) (int, int, nvml.Return) {
-			return 7, 5, nvml.SUCCESS
-		},
-		DeviceGetIndexFunc: func(nvml.Device) (int, nvml.Return) {
-			return 0, nvml.SUCCESS
-		},
-		DeviceGetMigModeFunc: func(nvml.Device) (int, int, nvml.Return) {
+	return mock
+}
+
+// GetMIGDeviceMock returns a mock of the MIG Device.
+func GetMIGDeviceMock(deviceIdx int, migDeviceIdx int, opts ...func(*nvmlmock.Device)) *nvmlmock.Device {
+	// Take the original mock and override only the relevant functions to make it a MIG device.
+	prepareMigDevice := func(d *nvmlmock.Device) {
+		// No MIG children
+		d.GetMigModeFunc = func() (int, int, nvml.Return) {
 			return nvml.DEVICE_MIG_DISABLE, 0, nvml.SUCCESS
-		},
-		DeviceGetComputeRunningProcessesFunc: func(nvml.Device) ([]nvml.ProcessInfo, nvml.Return) {
-			return DefaultProcessInfo, nvml.SUCCESS
-		},
-		DeviceGetMemoryInfoFunc: func(nvml.Device) (nvml.Memory, nvml.Return) {
-			return nvml.Memory{Total: DefaultTotalMemory, Free: 500}, nvml.SUCCESS
-		},
-		SystemGetDriverVersionFunc: func() (string, nvml.Return) {
-			return DefaultNvidiaDriverVersion, nvml.SUCCESS
-		},
+		}
+
+		// Change the UUID
+		d.GetUUIDFunc = func() (string, nvml.Return) {
+			return MIGChildrenUUIDs[deviceIdx][migDeviceIdx], nvml.SUCCESS
+		}
+
+		// Change the name
+		d.GetNameFunc = func() (string, nvml.Return) {
+			return "MIG " + DefaultGPUName, nvml.SUCCESS
+		}
+		d.IsMigDeviceHandleFunc = func() (bool, nvml.Return) {
+			return true, nvml.SUCCESS
+		}
+
+		// Override GetAttributesFunc for this specific MIG child to correctly distribute parent's resources.
+		d.GetAttributesFunc = func() (nvml.DeviceAttributes, nvml.Return) {
+			numMigChildrenForParent, ok := MIGChildrenPerDevice[deviceIdx]
+			if !ok || numMigChildrenForParent == 0 {
+				// Should not happen if MIGChildrenPerDevice is consistent for a MIG-enabled parent
+				// Return error
+				return nvml.DeviceAttributes{}, nvml.ERROR_NOT_SUPPORTED
+			}
+
+			// core count and total memory - equally distribute between all mig devices
```

Comment:
For simplicity the mock equally distributes resources across mig devices.
Real world scenarios may have specific allocations. While this is fine for the current tests, mentioning it in the comment could help clarify.